File: solutions/ruby/kindergarten-garden/1/kindergarten_garden.rb
class Garden
  CHILDREN = %w[alice bob charlie david eve fred ginny harriet ileana joseph kincaid larry]
  PLANT = {g: :grass, c: :clover, r: :radishes, v: :violets}
  
  def initialize(plants_str)
    plants = plants_str.split("\n").map(&:chars)
    num_students = plants[0].length/2
    rearranged_plants = (0...num_students).to_a.map{|index|index*2}.map{|index| plants[0][index..index+1]+plants[1][index..index+1]}.map{|plant_arr| plant_arr.map{|initial| PLANT[initial.downcase.to_sym]}}
    @kids_plants = Hash[CHILDREN[0...num_students].zip(rearranged_plants)]
  end

  def method_missing(child)
    @kids_plants[child.to_s]
  end
end

File: solutions/cpp/kindergarten-garden/1/kindergarten_garden.cpp
#include "kindergarten_garden.h"

namespace kindergarten_garden {
Plants initial_to_plant(char initial) {
    switch (initial) {
        case 'R': return Plants::radishes;
        case 'G': return Plants::grass;
        case 'C': return Plants::clover;
        case 'V': return Plants::violets;
    };
    throw (std::domain_error(""));
}
std::array<Plants, 4> plants(std::string rows, std::string name) {
    int offset = name.data()[0] - 'A';
    auto class_size = rows.length() / 2 + 1;
    auto initials = rows.data();
    std::array<Plants, 4> result;
    result[0] = initial_to_plant(initials[2*offset]);
    result[1] = initial_to_plant(initials[2*offset + 1]);
    result[2] = initial_to_plant(initials[class_size + 2*offset]);
    result[3] = initial_to_plant(initials[class_size + 2*offset + 1]);
    return result;
}

}  // namespace kindergarten_garden
